1364-integrate-tx-pause-safe-mode

sylvaincormier opened the pull request:
Date:   Thu Nov 7 08:10:27 2024 -0500

    Integrate SafeMode pallet into Shibuya runtime
    - Added pallet_tx_pause to runtime configuration.
    - Added `pallet_safe_mode` to runtime configuration.
    - Updated genesis configuration to include `safe_mode`.
    - Configured `SafeModeWhitelistedCalls` to define calls that bypass SafeMode restrictions.
    - Added parameter types for SafeMode configuration, including enter and extend durations, deposit amounts, and release delay.
    - Updated `construct_runtime!` macro to include SafeMode pallet at index

**Pull Request Summary**


**Check list**
- [ ] added or updated unit tests
- [ ] updated Astar official documentation
- [ ] added OnRuntimeUpgrade hook for precompile revert code registration
- [ ] added benchmarks & weights for any modified runtime logics.

ipapandinas commented on runtime/shibuya/src/lib.rs:
```diff
@@ -1522,7 +1526,13 @@ parameter_types! {
     type RuntimeEvent = RuntimeEvent;
     #[cfg(not(feature = "runtime-benchmarks"))]
-    type Migrations = ();
+    type Migrations = (
```
Those migrations have already been executed on Shibuya, and will be deleted once you have rebased the master branch

ipapandinas commented on runtime/shibuya/src/lib.rs:
```diff
@@ -1655,7 +1667,7 @@ pub type Executive = frame_executive::Executive<
 
 /// Unreleased migrations. Add new ones here:
-pub type Unreleased = ();
+pub type Unreleased = (cumulus_pallet_xcmp_queue::migration::v5::MigrateV4ToV5<Runtime>,);
```
same here, it has already been applied

ipapandinas commented on runtime/shibuya/src/lib.rs:
```diff
@@ -1608,6 +1618,8 @@ construct_runtime!(
         CollectiveProxy: pallet_collective_proxy = 109,
 
         MultiBlockMigrations: pallet_migrations = 120,
+        TxPause: pallet_tx_pause = 200,
```
A small gap is ok, we can go with SafeMode first:

```rs
SafeMode: pallet_safe_mode = 130,
TxPause: pallet_tx_pause = 131,
```

ipapandinas commented on runtime/shibuya/src/lib.rs:
```diff
@@ -1665,6 +1677,52 @@ type EventRecord = frame_system::EventRecord<
+    type EnterDepositAmount = EnterDepositAmount;
+    type ExtendDuration = ExtendDuration;
+    type ExtendDepositAmount = ExtendDepositAmount;
+    type ForceEnterOrigin = EnsureRootWithSuccess<AccountId, ConstU32<9>>;
```
A better configuration can be to use a special type for those privileges. You can create an `EmergencyAuthority` and `EmergencyAuthorityWithSuccess` in governance primitives:

```rs
// OpsMaintainer origins

pub type EmergencyAuthority = EitherOfDiverse<MoreThanHalfCouncil, EnsureRootOrAllTechnicalCommittee>;
pub type EmergencyAuthorityWithSuccess<Success = BlockNumberFor<Self>> = EnsureWithSuccess<EmergencyAuthority, AccountId, Success>;
```

and use them in the pallet_safe_mode config like this:

```rs
type ForceEnterOrigin = EmergencyAuthorityWithSuccess<ConstU32<2_400>>; // 2_400 blocks = 4 hours
type ForceExtendOrigin = EmergencyAuthorityWithSuccess<ConstU32<2_400>>; // 2_400 blocks = 4 hours
type ForceExitOrigin = EmergencyAuthority;
type ForceDepositOrigin = EmergencyAuthority;
```

ipapandinas commented on runtime/shibuya/src/lib.rs:
```diff
@@ -1665,6 +1677,52 @@ type EventRecord = frame_system::EventRecord<
+impl pallet_tx_pause::Config for Runtime {
+    type RuntimeCall = RuntimeCall;
+    type RuntimeEvent = RuntimeEvent;
+    type PauseOrigin = EnsureRoot<AccountId>; // Authority required to pause transactions.
```
Same here:

```rs
type PauseOrigin = EmergencyAuthority;
type UnpauseOrigin = EmergencyAuthority;
```

ipapandinas commented on runtime/shibuya/src/lib.rs:
```diff
@@ -1665,6 +1677,52 @@ type EventRecord = frame_system::EventRecord<
+    type PauseOrigin = EnsureRoot<AccountId>; // Authority required to pause transactions.
+    type UnpauseOrigin = EnsureRoot<AccountId>; // Authority required to unpause transactions.
+    type WhitelistedCalls = ();
+    type MaxNameLen = ConstU32<32>; // Maximum length of name, adjust as needed.
```
"256" is probably better here: `type MaxNameLen = ConstU32<256>`

ipapandinas commented on runtime/shibuya/src/lib.rs:
```diff
@@ -1665,6 +1677,52 @@ type EventRecord = frame_system::EventRecord<
+    type RuntimeEvent = RuntimeEvent;
+    type PauseOrigin = EnsureRoot<AccountId>; // Authority required to pause transactions.
+    type UnpauseOrigin = EnsureRoot<AccountId>; // Authority required to unpause transactions.
+    type WhitelistedCalls = ();
```
In the similar way you did for the pallet_safe_mode, can you create a `TxPauseWhitelistedCalls` type?

An example (but you can check if other pallets can be whitelisted):

```rs
/// Calls that can bypass the tx-pause pallet.
/// We always allow system calls and timestamp since it is required for block production
pub struct TxPauseWhitelistedCalls;
impl Contains<pallet_tx_pause::RuntimeCallNameOf<Runtime>> for TxPauseWhitelistedCalls {
	fn contains(full_name: &pallet_tx_pause::RuntimeCallNameOf<Runtime>) -> bool {
		matches!(full_name.0.as_slice(), b"Sudo" | b"System" | b"Timestamp" | b"TxPause")
	}
}
```

ipapandinas commented on runtime/shiden/src/lib.rs:
```diff
@@ -1194,7 +1194,13 @@ parameter_types! {
     type RuntimeEvent = RuntimeEvent;
     #[cfg(not(feature = "runtime-benchmarks"))]
-    type Migrations = ();
+    type Migrations = (
```
Same as above

ipapandinas commented on runtime/shiden/src/lib.rs:
```diff
@@ -1333,7 +1339,11 @@ parameter_types! {
 
 /// Unreleased migrations. Add new ones here:
-pub type Unreleased = ();
+pub type Unreleased = (
```
Same as above

ipapandinas commented on runtime/shibuya/src/lib.rs:
```diff
@@ -1665,6 +1677,52 @@ type EventRecord = frame_system::EventRecord<
+impl Contains<RuntimeCall> for SafeModeWhitelistedCalls {
+    fn contains(call: &RuntimeCall) -> bool {
+        match call {
+            RuntimeCall::System(_) | RuntimeCall::SafeMode(_) | RuntimeCall::TxPause(_) => true,
```
You can add `Sudo` too

ipapandinas reviewed (changes requested):
This first version is promising—good job! I've left comments on the config parameters and the master branch rebase.

Additionally, from my analysis of these pallets, it appears you forgot to extend the `BaseCallFilter` type in the `frame_system` config to include whitelisted calls. You can define a new type alias, `SafeModeTxPauseFilter`, and use it as follows:

```rs
type SafeModeTxPauseFilter = InsideBoth<SafeModeWhitelistedCalls, TxPauseWhitelistedCalls>;

type BaseCallFilter = InsideBoth<BaseFilter, SafeModeTxPauseFilter>;
```

sylvaincormier commented:
@ipapandinas, thank you for taking the time to provide such thorough feedback and detailed examples! I've addressed all the points you raised:

    Updated BaseCallFilter to incorporate both SafeModeWhitelistedCalls and TxPauseWhitelistedCalls.
    Created TxPauseWhitelistedCalls to ensure critical calls like Sudo, System, Timestamp, and TxPause are whitelisted.
    Extended SafeModeWhitelistedCalls to include Sudo.
    Reordered the SafeMode and TxPause pallets as suggested.
    Increased MaxNameLen to 256.

I also rebased onto the latest master as you recommended. Let me know if there's anything else you need me to adjust. I appreciate your guidance throughout this process!

ipapandinas commented:
@sylvaincormier Thank you for your work on this contribution! For now, we’ll close this pull request, but we appreciate your effort and may revisit these ideas in the future. Thanks again for your understanding!

sylvaincormier commented:
Hi Ipapandinas.
I was not done!
I just finished this:
Author: Silvereau <sylvaincormier@protonmail.com>
Date:   Wed Nov 13 06:02:18 2024 -0500

    feat(runtime): add EmergencyAuthority origin types
    
    Define origin types for emergency operations:
    - MoreThanHalfCouncil for council majority actions
    - EmergencyAuthority combining council and technical committee control
    - EmergencyAuthorityWithSuccess for time-sensitive operations
    
    These types enable proper governance control over emergency features.
    
    Why is this pull request closed and not integrated into main?
